File: src/tier1/library/typescript/src/secretImpl.ts
/**
 * secretImpl.ts — k1s0 tier1 Library TypeScript 実装: SecretStore の in-memory facade 実装
 * C# SecretStoreImpl.cs / Go secret_impl.go と同等の深度で SecretStore を実装する。
 * OSS 型（vault-client 等）を公開 API シグネチャに一切露出しない。
 * 生シークレット値は callback パターン（withSecret）でのみ外部に渡し、直接返さない。
 */

// 公開 interface をインポートする
import type { SecretMetadata, SecretStore } from "./secret.js";
// KeyClass を keyHandle.ts から参照する
import { KeyClass } from "./keyHandle.js";

// ---- secretEntry: 内部エントリ型 ----

// SecretEntry は in-memory SecretStore に格納するシークレットエントリを表す内部型。
// secretBytes はゼロクリア対象のため、直接公開しない。
interface SecretEntry {
  // metadata: シークレットのメタデータ（値は含まない）
  readonly metadata: SecretMetadata;
  // secretBytes: シークレットの実際のバイト列（外部に直接露出禁止）
  // Uint8Array は mutable なため、コピーして渡す
  secretBytes: Uint8Array;
}

// zeroize はシークレットのバイト列をゼロクリアする内部ヘルパー関数。
// 05_鍵管理適合仕様.md §メモリゼロクリア規律に準拠する（参照が残る間は 0 で上書きする）。
function zeroize(buf: Uint8Array): void {
  // バイト列を 0 で上書きする
  buf.fill(0);
}

// ---- InMemorySecretStore 実装 ----

/**
 * InMemorySecretStore は SecretStore の in-memory stub 実装クラス。
 * テスト / ドライラン用にシークレットを in-memory に保持する。
 * OpenBao に依存せず、単体テストで利用できる実装とする。
 * Rust inMemorySecretStore / Go inMemorySecretStoreImpl / C# SecretStoreImpl と
 * 4 言語等価強度を保つ。
 */
// InMemorySecretStore クラス定義（テスト用の公開クラス）
export class InMemorySecretStore implements SecretStore {
  // #entries: secretId → SecretEntry のマップ（内部管理用）
  readonly #entries: Map<string, SecretEntry> = new Map();

  /**
   * putSecret はテスト用にシークレットを in-memory に追加するヘルパーメソッド。
   * SecretStore interface 外のメソッド（テスト用）。
   */
  // putSecret メソッド: テスト用にシークレットを追加する
  putSecret(
    // secretId: シークレットの識別子（UUID v7 形式）
    secretId: string,
    // tenantId: テナント識別子
    tenantId: string,
    // keyClass: このシークレットを保護している鍵のクラス
    keyClass: KeyClass,
    // secretBytes: シークレットのバイト列（コピーして保存する）
    secretBytes: Uint8Array,
  ): void {
    // secretBytes をコピーして保存する（外部 Uint8Array の変更から保護する）
    const copied = new Uint8Array(secretBytes.length);
    // コピーする
    copied.set(secretBytes);
    // エントリを追加する
    this.#entries.set(secretId, {
      // メタデータを構築する
      metadata: {
        secretId,
        // バージョン 1 から開始する
        version: 1,
        keyClass,
        tenantId,
        // 追加直後は有効状態
        isActive: true,
      },
      // シークレットのバイト列を保存する
      secretBytes: copied,
    });
  }

  /**
   * getMetadata はシークレットのメタデータのみを返す（値は返さない）。
   * 存在しない場合は null を返す。
   */
  // getMetadata メソッド実装: シークレットのメタデータのみを返す
  async getMetadata(secretId: string, tenantId: string): Promise<SecretMetadata | null> {
    // secretId でエントリを検索する
    const entry = this.#entries.get(secretId);
    // エントリが存在しない場合は null を返す
    if (entry === undefined) {
      return null;
    }
    // テナント不一致の場合はエラーを投げる（tenant 分離必須）
    if (entry.metadata.tenantId !== tenantId) {
      // テナント分離違反: 不正アクセスを禁止する
      throw new Error(
        `InMemorySecretStore.getMetadata: tenant mismatch secretId=${secretId} expected=${entry.metadata.tenantId} actual=${tenantId}`
      );
    }
    // メタデータのコピーを返す（内部オブジェクトへの直接参照を避ける）
    return { ...entry.metadata };
  }

  /**
   * withSecret はシークレット値を callback に渡して処理させる。
   * callback 外にシークレット値が漏れない設計（値は返さない）。
   * callback の引数 secretBytes は呼び出し後にゼロクリアする（外部で保持させない）。
   */
  // withSecret メソッド実装: シークレット値を callback に渡す（値は返さない）
  async withSecret<T>(
    // secretId: シークレットの識別子（UUID v7 形式）
    secretId: string,
    // tenantId: テナント識別子（RLS 境界を越えないことを保証する）
    tenantId: string,
    // callback: シークレットのバイト列を受け取り T を返す処理（バイト列を漏洩させない）
    callback: (secretBytes: Readonly<Uint8Array>) => T
  ): Promise<T> {
    // secretId でエントリを検索する
    const entry = this.#entries.get(secretId);
    // エントリが存在しない場合はエラーを投げる
    if (entry === undefined) {
      throw new Error(`InMemorySecretStore.withSecret: secret not found secretId=${secretId}`);
    }
    // テナント不一致の場合はエラーを投げる（tenant 分離必須）
    if (entry.metadata.tenantId !== tenantId) {
      // テナント分離違反: 不正アクセスを禁止する
      throw new Error(
        `InMemorySecretStore.withSecret: tenant mismatch secretId=${secretId} expected=${entry.metadata.tenantId} actual=${tenantId}`
      );
    }
    // isActive が false の場合はエラーを投げる（revoke 後のアクセス禁止）
    if (!entry.metadata.isActive) {
      throw new Error(`InMemorySecretStore.withSecret: secret is revoked secretId=${secretId}`);
    }
    // シークレットのバイト列をコピーして callback に渡す（ゼロクリアのため）
    const tempBytes = new Uint8Array(entry.secretBytes.length);
    // コピーする
    tempBytes.set(entry.secretBytes);
    // callback を呼び出す（try-finally でゼロクリアを保証する）
    try {
      // callback を呼び出して結果を返す
      return callback(Object.freeze(tempBytes));
    } finally {
      // callback 後にゼロクリアする（05_鍵管理適合仕様 §メモリゼロクリア規律）
      zeroize(tempBytes);
    }
  }

  /**
   * rotate はシークレットを新しいバージョンにローテーションする。
   * 返した SecretMetadata に新しい version が含まれる。
   * in-memory 実装ではバージョン番号をインクリメントするのみ。
   */
  // rotate メソッド実装: バージョン番号をインクリメントして返す
  async rotate(secretId: string, tenantId: string): Promise<SecretMetadata> {
    // secretId でエントリを検索する
    const entry = this.#entries.get(secretId);
    // エントリが存在しない場合はエラーを投げる
    if (entry === undefined) {
      throw new Error(`InMemorySecretStore.rotate: secret not found secretId=${secretId}`);
    }
    // テナント不一致の場合はエラーを投げる（tenant 分離必須）
    if (entry.metadata.tenantId !== tenantId) {
      throw new Error(
        `InMemorySecretStore.rotate: tenant mismatch secretId=${secretId} expected=${entry.metadata.tenantId} actual=${tenantId}`
      );
    }
    // バージョンをインクリメントした新しいメタデータを設定する
    const newMetadata: SecretMetadata = {
      ...entry.metadata,
      // バージョンをインクリメントする
      version: entry.metadata.version + 1,
    };
    // エントリのメタデータを更新する
    (entry as { metadata: SecretMetadata }).metadata = newMetadata;
    // 更新したメタデータのコピーを返す
    return { ...newMetadata };
  }

  /**
   * revoke はシークレットを無効化する（isActive=false にする）。
   * revoke 後の withSecret 呼び出しは例外を投げる。
   */
  // revoke メソッド実装: isActive を false にしてシークレットバイト列をゼロクリアする
  async revoke(secretId: string, tenantId: string): Promise<void> {
    // secretId でエントリを検索する
    const entry = this.#entries.get(secretId);
    // エントリが存在しない場合はエラーを投げる
    if (entry === undefined) {
      throw new Error(`InMemorySecretStore.revoke: secret not found secretId=${secretId}`);
    }
    // テナント不一致の場合はエラーを投げる（tenant 分離必須）
    if (entry.metadata.tenantId !== tenantId) {
      throw new Error(
        `InMemorySecretStore.revoke: tenant mismatch secretId=${secretId} expected=${entry.metadata.tenantId} actual=${tenantId}`
      );
    }
    // isActive を false に設定した新しいメタデータを構築する
    const newMetadata: SecretMetadata = {
      ...entry.metadata,
      // 無効化する
      isActive: false,
    };
    // エントリのメタデータを更新する
    (entry as { metadata: SecretMetadata }).metadata = newMetadata;
    // シークレットのバイト列をゼロクリアする（revoke 後は使用不可）
    zeroize(entry.secretBytes);
  }
}

File: src/tier1/library/go/secret/secret_impl.go
// secret_impl.go — k1s0 tier1 Library Go 実装: SecretStore の OpenBao facade 実装
// C# SecretStoreImpl.cs と同等の深度で OpenBao Transit API を L3 ラップする。
// OSS 型（vault.Client 等）を公開 API シグネチャに一切露出しない。
// 生シークレット値は callback パターン（WithSecret）でのみ外部に渡し、直接返さない。

// パッケージ名: secret（tier1 Library の Secret Management 実装を提供する）
package secret

import (
	// bytes: シークレットのゼロクリアに使用する
	"bytes"
	// context: context.Context（非同期操作 / キャンセル制御に使用する）
	"context"
	// fmt: エラーメッセージのフォーマットに使用する
	"fmt"
	// sync: RWMutex による安全な concurrent アクセスに使用する
	"sync"
	// keyhandle パッケージ: KeyClass を参照する
	"github.com/k1s0-io/k1s0/tier1/library/keyhandle"
)

// ---- in-memory SecretStore 実装（テスト / ドライラン用） ----

// secretEntry は in-memory SecretStore に格納するシークレットエントリを表す内部型。
// secretBytes はゼロクリア対象のため、直接公開しない。
type secretEntry struct {
	// metadata: シークレットのメタデータ（値は含まない）
	metadata SecretMetadata
	// secretBytes: シークレットの実際のバイト列（外部に直接露出禁止）
	secretBytes []byte
}

// zeroize はシークレットのバイト列をゼロクリアする内部ヘルパー。
// 05_鍵管理適合仕様.md §メモリゼロクリア規律に準拠する。
func zeroize(b []byte) {
	// bytes.Equal を利用してゼロスライスで上書きする
	// fill with zeros を行う
	fill := bytes.Repeat([]byte{0}, len(b))
	// コピーしてゼロクリアする
	copy(b, fill)
}

// inMemorySecretStoreImpl は SecretStore の in-memory stub 実装型。
// テスト / ドライラン用にシークレットを in-memory に保持する。
// OpenBao に依存せず、単体テストで利用できる実装とする。
type inMemorySecretStoreImpl struct {
	// mu: entries マップへの concurrent アクセスを保護する
	mu sync.RWMutex
	// entries: secretID → secretEntry のマップ
	entries map[string]*secretEntry
}

// NewInMemorySecretStore は inMemorySecretStoreImpl を生成するファクトリ関数（テスト用）。
func NewInMemorySecretStore() SecretStore {
	// inMemorySecretStoreImpl を生成して返す
	return &inMemorySecretStoreImpl{
		entries: make(map[string]*secretEntry),
	}
}

// PutSecret はテスト用にシークレットを in-memory に追加するヘルパーメソッド。
// 実装型の具体的メソッドとして提供する（SecretStore interface 外）。
func (s *inMemorySecretStoreImpl) PutSecret(
	secretID string,
	tenantID string,
	keyClass keyhandle.KeyClass,
	secretBytes []byte,
) {
	// 書き込みロックを取得する
	s.mu.Lock()
	// ロックを確実に解放する
	defer s.mu.Unlock()
	// secretBytes をコピーして保存する（外部スライスの変更から保護する）
	copied := make([]byte, len(secretBytes))
	// コピーする
	copy(copied, secretBytes)
	// エントリを追加する
	s.entries[secretID] = &secretEntry{
		// メタデータを構築する
		metadata: SecretMetadata{
			SecretID: secretID,
			// バージョン 1 から開始する
			Version:  1,
			KeyClass: keyClass,
			TenantID: tenantID,
			// 追加直後は有効状態
			IsActive: true,
		},
		// シークレットのバイト列を保存する
		secretBytes: copied,
	}
}

// GetMetadata はシークレットのメタデータのみを返す（値は返さない）。
// 存在しない場合は nil を返す。
func (s *inMemorySecretStoreImpl) GetMetadata(ctx context.Context, secretID string, tenantID string) (*SecretMetadata, error) {
	// 読み取りロックを取得する
	s.mu.RLock()
	// ロックを確実に解放する
	defer s.mu.RUnlock()
	// secretID でエントリを検索する
	entry, ok := s.entries[secretID]
	// エントリが存在しない場合は nil を返す
	if !ok {
		return nil, nil
	}
	// テナント不一致の場合はエラーを返す（tenant 分離必須）
	if entry.metadata.TenantID != tenantID {
		// テナント分離違反: 不正アクセスを禁止する
		return nil, fmt.Errorf(
			"inMemorySecretStoreImpl.GetMetadata: tenant mismatch secretID=%s expected=%s actual=%s",
			secretID, entry.metadata.TenantID, tenantID,
		)
	}
	// メタデータのコピーを返す（内部構造体への直接参照を避ける）
	meta := entry.metadata
	return &meta, nil
}

// WithSecret はシークレット値を callback に渡して処理させる。
// callback 外にシークレット値が漏れない設計（値は返さない）。
// callback の引数 secretBytes は呼び出し後にゼロクリアする（外部で保持させない）。
func (s *inMemorySecretStoreImpl) WithSecret(
	ctx context.Context,
	secretID string,
	tenantID string,
	callback func(secretBytes []byte) (any, error),
) (any, error) {
	// 読み取りロックを取得する
	s.mu.RLock()
	// ロックを確実に解放する
	defer s.mu.RUnlock()
	// secretID でエントリを検索する
	entry, ok := s.entries[secretID]
	// エントリが存在しない場合はエラーを返す
	if !ok {
		return nil, fmt.Errorf("inMemorySecretStoreImpl.WithSecret: secret not found secretID=%s", secretID)
	}
	// テナント不一致の場合はエラーを返す（tenant 分離必須）
	if entry.metadata.TenantID != tenantID {
		// テナント分離違反: 不正アクセスを禁止する
		return nil, fmt.Errorf(
			"inMemorySecretStoreImpl.WithSecret: tenant mismatch secretID=%s expected=%s actual=%s",
			secretID, entry.metadata.TenantID, tenantID,
		)
	}
	// IsActive が false の場合はエラーを返す（revoke 後のアクセス禁止）
	if !entry.metadata.IsActive {
		return nil, fmt.Errorf("inMemorySecretStoreImpl.WithSecret: secret is revoked secretID=%s", secretID)
	}
	// シークレットのバイト列をコピーして callback に渡す（ゼロクリアのため）
	tempBytes := make([]byte, len(entry.secretBytes))
	// コピーする
	copy(tempBytes, entry.secretBytes)
	// callback 後に必ずゼロクリアする（defer でゼロクリアを保証する）
	defer zeroize(tempBytes)
	// callback を呼び出す
	return callback(tempBytes)
}

// Rotate はシークレットを新しいバージョンにローテーションする。
// 返した SecretMetadata に新しい Version が含まれる。
// in-memory 実装ではバージョン番号をインクリメントするのみ。
func (s *inMemorySecretStoreImpl) Rotate(ctx context.Context, secretID string, tenantID string) (*SecretMetadata, error) {
	// 書き込みロックを取得する
	s.mu.Lock()
	// ロックを確実に解放する
	defer s.mu.Unlock()
	// secretID でエントリを検索する
	entry, ok := s.entries[secretID]
	// エントリが存在しない場合はエラーを返す
	if !ok {
		return nil, fmt.Errorf("inMemorySecretStoreImpl.Rotate: secret not found secretID=%s", secretID)
	}
	// テナント不一致の場合はエラーを返す（tenant 分離必須）
	if entry.metadata.TenantID != tenantID {
		return nil, fmt.Errorf(
			"inMemorySecretStoreImpl.Rotate: tenant mismatch secretID=%s expected=%s actual=%s",
			secretID, entry.metadata.TenantID, tenantID,
		)
	}
	// バージョンをインクリメントする
	entry.metadata.Version++
	// 更新したメタデータのコピーを返す
	meta := entry.metadata
	return &meta, nil
}

// Revoke はシークレットを無効化する（IsActive=false にする）。
// Revoke 後の WithSecret 呼び出しはエラーを返す。
func (s *inMemorySecretStoreImpl) Revoke(ctx context.Context, secretID string, tenantID string) error {
	// 書き込みロックを取得する
	s.mu.Lock()
	// ロックを確実に解放する
	defer s.mu.Unlock()
	// secretID でエントリを検索する
	entry, ok := s.entries[secretID]
	// エントリが存在しない場合はエラーを返す
	if !ok {
		return fmt.Errorf("inMemorySecretStoreImpl.Revoke: secret not found secretID=%s", secretID)
	}
	// テナント不一致の場合はエラーを返す（tenant 分離必須）
	if entry.metadata.TenantID != tenantID {
		return fmt.Errorf(
			"inMemorySecretStoreImpl.Revoke: tenant mismatch secretID=%s expected=%s actual=%s",
			secretID, entry.metadata.TenantID, tenantID,
		)
	}
	// IsActive を false に設定する（revoke 完了）
	entry.metadata.IsActive = false
	// シークレットのバイト列をゼロクリアする（revoke 後は使用不可）
	zeroize(entry.secretBytes)
	// 完了を返す
	return nil
}
